File: pythonbindings/examples/NormalGG/NormalGG.py
import sys
import os
from pathlib import Path

lib_dir1 = Path(__file__).resolve().parents[3] / "lib" / "libOpenPhase.so"

link1 = Path("libOpenPhase.so")
if not link1.exists():
    os.symlink(lib_dir1, link1)
    print(f"Created symlink: {link1} → {lib_dir1}")
else:
    print(f"Symlink already exists: {link1}")
    
sys.path.append(str(Path(__file__).resolve().parents[2]))

from OpenPhase import (
    Settings,
    RunTimeControl,
    PhaseField,
    DoubleObstacle,
    InterfaceProperties,
    BoundaryConditions,
    Initializations,
    TimeInfo,
)
# Optional (only if available in bindings)
try:
    from OpenPhase import MicrostructureAnalysis, ConsoleOutput
except ImportError:
    MicrostructureAnalysis = None
    ConsoleOutput = None


def main():
    # --- Handle command line arguments ---
    if len(sys.argv) > 1:
        input_file = sys.argv[1]
    else:
        print("No Inputfile provided, trying to use default ProjectInput.opi")
        input_file = "ProjectInput.opi"

    # --- Initialize core simulation objects ---
    OPSettings = Settings(input_file)

    RTC = RunTimeControl(OPSettings, input_file)
    Phi = PhaseField(OPSettings, input_file)
    DO = DoubleObstacle(OPSettings, input_file)
    IP = InterfaceProperties(OPSettings, input_file)
    BC = BoundaryConditions(OPSettings, input_file)
    Timer = TimeInfo(OPSettings, "Execution Time Statistics")

    # --- Generate initial microstructure (Voronoi) ---
    number_of_grains = 200
    GrainsPhase = 0
    Initializations.VoronoiTessellation(Phi, BC, number_of_grains, GrainsPhase)

    print("Entering the Time Loop!!!")

    # --- Main time-stepping loop ---
    t = RTC.StartTimeStep
    while t <= RTC.MaxTimeStep:
        Timer.SetStart()

        IP.Set(Phi, BC)
        Timer.SetTimeStamp("IP.Set()")

        DO.CalculatePhaseFieldIncrements(Phi, IP)
        Timer.SetTimeStamp("CalculatePhaseFieldIncrements")

        Phi.NormalizeIncrements(BC, RTC.dt)
        Timer.SetTimeStamp("NormalizeIncrements")

        Phi.MergeIncrements(BC, RTC.dt)
        Timer.SetTimeStamp("MergeIncrements")

        # --- Output to VTK ---
        if RTC.WriteVTK():
            Phi.WriteVTK(OPSettings, RTC.TimeStep)
            if MicrostructureAnalysis:
                MicrostructureAnalysis.WriteGrainsStatistics(Phi, RTC.TimeStep)

        # --- Output raw data ---
        if RTC.WriteRawData():
            Phi.Write(OPSettings, RTC.TimeStep)

        # --- Output to screen ---
        if RTC.WriteToScreen():
            I_En = DO.AverageEnergyDensity(Phi, IP)
            if ConsoleOutput:
                msg = ConsoleOutput.GetStandard("Interface energy density", I_En)
                ConsoleOutput.WriteTimeStep(RTC, msg)
            else:
                print(f"[t={RTC.TimeStep}] Interface energy density: {I_En:.6e}")

            Timer.PrintWallClockSummary()

        # --- Increment timestep ---
        RTC.IncrementTimeStep()
        t = RTC.TimeStep

    print("Simulation finished successfully.")


if __name__ == "__main__":
    main()

File: examples/NormalGG/NormalGG.cpp
/*
 *   This file is part of the OpenPhase (R) software library.
 *  
 *  Copyright (c) 2009-2025 Ruhr-Universitaet Bochum,
 *                Universitaetsstrasse 150, D-44801 Bochum, Germany
 *            AND 2018-2025 OpenPhase Solutions GmbH,
 *                Universitaetsstrasse 136, D-44799 Bochum, Germany.
 *  
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *     
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *  
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.

 *   File created :   2011
 *   Main contributors :  Reza Darvishi Kamachali; Oleg Shchyglo; Hesham Salama
 *
 */

#include "Settings.h"
#include "RunTimeControl.h"
#include "DoubleObstacle.h"
#include "PhaseField.h"
#include "Initializations.h"
#include "BoundaryConditions.h"
#include "InterfaceProperties.h"
#include "Tools/TimeInfo.h"
#include "Tools/MicrostructureAnalysis.h"

using namespace openphase;

int main(int argc, char *argv[])
{
#ifdef MPI_PARALLEL
    int provided = 0;
    OP_MPI_Init_thread(&argc, &argv, OP_MPI_THREAD_FUNNELED, &provided);
    OP_MPI_Comm_rank(OP_MPI_COMM_WORLD, &MPI_RANK);
    OP_MPI_Comm_size(OP_MPI_COMM_WORLD, &MPI_SIZE);
    {
#endif
	std::string InputFile;
    if(argc > 1)
    {
        InputFile = argv[1];
    }
    else
    {
        std::cerr << "No Inputfile provided, trying to use default ProjectInput.opi" << std::endl;
        InputFile = "ProjectInput.opi";
    }
    Settings                    OPSettings;
    OPSettings.ReadInput(InputFile);

    RunTimeControl              RTC(OPSettings, InputFile);
    PhaseField                  Phi(OPSettings, InputFile);
    DoubleObstacle              DO(OPSettings, InputFile);
    InterfaceProperties         IP(OPSettings, InputFile);
    BoundaryConditions          BC(OPSettings, InputFile);
    TimeInfo                    Timer(OPSettings, "Execution Time Statistics");

    //generating initial grain structure using Voronoi algorithm
    int number_of_grains = 200;
    size_t GrainsPhase = 0;
    Initializations::VoronoiTessellation(Phi, BC, number_of_grains, GrainsPhase);

    std::cout << "Entering the Time Loop!!!" << std::endl;
    for(RTC.tStep = RTC.tStart; RTC.tStep <= RTC.nSteps; RTC.IncrementTimeStep())
    {
        Timer.SetStart();
        IP.Set(Phi, BC);
        Timer.SetTimeStamp("IP.Set()");
        DO.CalculatePhaseFieldIncrements(Phi, IP);
        Timer.SetTimeStamp("CalculatePhaseFieldIncrements");
        Phi.NormalizeIncrements(BC, RTC.dt);
        Timer.SetTimeStamp("NormalizeIncrements");
        Phi.MergeIncrements(BC, RTC.dt);
        Timer.SetTimeStamp("MergeIncrements");

        /// Output to VTK file
        if (RTC.WriteVTK())
        {
            Phi.WriteVTK(OPSettings, RTC.tStep);
            MicrostructureAnalysis::WriteGrainsStatistics(Phi,RTC.tStep);
        }
        /// Output raw data
        if (RTC.WriteRawData())
        {
            Phi.Write(OPSettings, RTC.tStep);
        }
        /// Output to screen
        if (RTC.WriteToScreen())
        {
            double I_En = DO.AverageEnergyDensity(Phi, IP);
            std::string message = ConsoleOutput::GetStandard("Interface energy density", I_En);
            ConsoleOutput::WriteTimeStep(RTC, message);
            Timer.PrintWallClockSummary();
        }
    }
#ifdef MPI_PARALLEL
    }
    OP_MPI_Finalize ();
#endif
    return EXIT_SUCCESS;
}
